internal/controller/atlasnetworkpeering/atlasnetworkpeering_controller.go

```diff
@@ -0,0 +1,433 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *AtlasNetworkPeeringReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Infow("-> Starting AtlasNetworkPeering reconciliation")
+
+	akoNetworkPeering := akov2.AtlasNetworkPeering{}
+	result := customresource.PrepareResource(ctx, r.Client, req, &akoNetworkPeering, r.Log)
+	if !result.IsOk() {
+		return result.ReconcileResult(), errors.New(result.GetMessage())
+	}
+
+	typeName := reflect.TypeOf(akoNetworkPeering).Name()
+	if customresource.ReconciliationShouldBeSkipped(&akoNetworkPeering) {
+		return r.Skip(ctx, typeName, &akoNetworkPeering, &akoNetworkPeering.Spec)
+	}
+
+	conditions := api.InitCondition(&akoNetworkPeering, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(r.Log, conditions, ctx, &akoNetworkPeering)
+	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, &akoNetworkPeering)
+
+	isValid := customresource.ValidateResourceVersion(workflowCtx, &akoNetworkPeering, r.Log)
+	if !isValid.IsOk() {
+		return r.Invalidate(typeName, isValid)
+	}
+
+	if !r.AtlasProvider.IsResourceSupported(&akoNetworkPeering) {
+		return r.Unsupport(workflowCtx, typeName)
+	}
+
+	credentials, err := r.ResolveCredentials(ctx, &akoNetworkPeering)
+	if err != nil {
+		return r.terminate(workflowCtx, &akoNetworkPeering, workflow.AtlasAPIAccessNotConfigured, err), nil
+	}
+	sdkClient, orgID, err := r.AtlasProvider.SdkClient(ctx, credentials, r.Log)
+	if err != nil {
+		return r.terminate(workflowCtx, &akoNetworkPeering, workflow.AtlasAPIAccessNotConfigured, err), nil
+	}
+	project, err := r.ResolveProject(ctx, sdkClient, &akoNetworkPeering, orgID)
+	if err != nil {
+		return r.terminate(workflowCtx, &akoNetworkPeering, workflow.AtlasAPIAccessNotConfigured, err), nil
+	}
+	return r.handle(&reconcileRequest{
+		workflowCtx:    workflowCtx,
+		service:        networkpeering.NewNetworkPeeringService(sdkClient.NetworkPeeringApi),
+		projectID:      project.ID,
+		networkPeering: &akoNetworkPeering,
+	}), nil
+}
+
+func (r *AtlasNetworkPeeringReconciler) handle(req *reconcileRequest) ctrl.Result {
+	r.Log.Infow("handling network peering reconcile request",
+		"service set", (req.service != nil), "projectID", req.projectID, "networkPeering", req.networkPeering)
+	var atlasPeer *networkpeering.NetworkPeer
+	if req.networkPeering.Status.ID != "" {
+		peer, err := req.service.GetPeer(req.workflowCtx.Context, req.projectID, req.networkPeering.Status.ID)
+		if err != nil && !errors.Is(err, networkpeering.ErrNotFound) {
+			return r.terminate(req.workflowCtx, req.networkPeering, workflow.Internal, err)
+		}
+		atlasPeer = peer
+	}
+	inAtlas := atlasPeer != nil
+	deleted := req.networkPeering.DeletionTimestamp != nil
```

Comment:
blocking: this doesn't catch the case of deletion timestamp being zero. Please use the `IsZero()` method which asserts both nil and zero time.